executor/table_reader.go

```diff
@@ -218,9 +237,26 @@ func (e *TableReaderExecutor) Next(ctx context.Context, req *chunk.Chunk) error
 		return err
 	}
 
+	// When 'select ... for update' work on a partitioned table, the table reader should
+	// add the partition ID as an extra column. The SelectLockExec need this information
+	// to construct the lock key.
+	physicalID := getPhysicalTableID(e.table)
+	if e.extraPIDColumnIndex.valid() {
+		fillExtraPIDColumn(req, e.extraPIDColumnIndex.value(), physicalID)
+	}
```

Comment:
Should we also do this in `IndexReader` and `IndexLookUp`?

Reply:
This might be caused by the extra partition column I added.
Since the schema columns are now more than just the index column, the index read plan is not generated.

https://github.com/pingcap/tidb/pull/21148/files#diff-4c4a5d9c358008fc0d3e898b87a14692668d4b60d949b2796d8b6ebc2b890720R2679

Reply:
How about merge this first and improve here later? 
If there is no `IndexReader` generated, there is no problem ... 
The test code cover the necessary cases and will prevent regression.

Reply:
OK for me.